### PR TITLE
move state extraction for CaptureMapDataset

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Simplifies the `_capture_metadata_collate` custom collate function when  capturing of state dict in the CaptureMapDataset.
This change was proposed by @tchaton in offline discussion.

Part of #9130 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/auto_restart.py [hunk 1]
```diff
@@ -231,8 +231,7 @@ def __getitem__(self, item) -> Tuple[Any, Dict[int, Dict]]:
             self._cached_state_dict = None
 
         data = self.dataset[item]
-        state_dict = self._state_dict()
-        return data, state_dict
+        return data
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:234]: nit:

```suggestion
return self.dataset[item]
```

File: pytorch_lightning/utilities/auto_restart.py [hunk 2]
```diff
@@ -464,17 +463,11 @@ def _capture_metadata_collate(samples: List, dataset: Dataset, default_collate:
...
-
-    elif isinstance(dataset, CaptureMapDataset):
-        samples, states = zip(*samples)
-        data = default_collate(samples)
-        metadata = states[-1]
-    else:
+    if not isinstance(dataset, (CaptureIterableDataset, CaptureMapDataset)):
         return default_collate(samples)
 
+    data = default_collate(samples)
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:469]: nit:

```suggestion
    data = default_collate(samples)
    if not isinstance(dataset, (CaptureIterableDataset, CaptureMapDataset)):
        return data
```

**Review by carmocca — APPROVED** [review]
